File: src/main/transcode.ts
import { spawn } from "child_process";
import { open } from "fs/promises";
import { Readable } from "stream";
import ffmpegStatic from "ffmpeg-static";
import { NATIVE_FORMATS } from "./audio-formats";

// Packaged app: ffmpeg binary lives in app.asar.unpacked, not app.asar
const FFMPEG_PATH =
  ffmpegStatic?.replace("app.asar", "app.asar.unpacked") ?? "ffmpeg";

const MP4_FORMATS = new Set(["m4a", "mp4"]);
const MOOV_SCAN_BYTES = 64 * 1024;

// Transcode output format: PCM s16le stereo @ 44100Hz
const SAMPLE_RATE = 44100;
const CHANNELS = 2;
const BYTES_PER_SAMPLE = 2;
const BYTES_PER_FRAME = CHANNELS * BYTES_PER_SAMPLE;
const PCM_BYTES_PER_SECOND = SAMPLE_RATE * BYTES_PER_FRAME;
export const WAV_HEADER_SIZE = 44;

export function needsTranscode(format: string): boolean {
  return !NATIVE_FORMATS.has(format.toLowerCase());
}

// MP4 files are streamable by Chromium only when the 'moov' atom precedes
// 'mdat' (faststart layout). Files with moov-at-end fail mid-playback even
// with byte-correct Range responses, so we transcode those.
export async function isMp4FastStart(filePath: string): Promise<boolean> {
  let fd;
  try {
    fd = await open(filePath, "r");
  } catch {
    return false;
  }
  try {
    const buf = Buffer.alloc(MOOV_SCAN_BYTES);
    const { bytesRead } = await fd.read(buf, 0, buf.length, 0);
    let offset = 0;
    while (offset + 8 <= bytesRead) {
      const size = buf.readUInt32BE(offset);
      const type = buf.toString("ascii", offset + 4, offset + 8);
      if (type === "moov") return true;
      if (type === "mdat") return false;
      let advance: number;
      if (size === 0) return false; // atom extends to EOF — moov not seen
      if (size === 1) {
        if (offset + 16 > bytesRead) return false;
        const hi = buf.readUInt32BE(offset + 8);
        const lo = buf.readUInt32BE(offset + 12);
        advance = hi * 0x1_0000_0000 + lo;
      } else {
        advance = size;
      }
      if (advance < 8) return false; // malformed
      offset += advance;
    }
    return false;
  } finally {
    await fd.close();
  }
}

export async function shouldTranscode(
  format: string,
  filePath: string,
): Promise<boolean> {
  const fmt = format.toLowerCase();
  if (MP4_FORMATS.has(fmt)) return !(await isMp4FastStart(filePath));
  return needsTranscode(fmt);
}

export function pcmBytesForDuration(durationSeconds: number): number {
  // Round down to whole sample frames (4 bytes for s16le stereo).
  return (
    Math.floor((durationSeconds * PCM_BYTES_PER_SECOND) / BYTES_PER_FRAME) *
    BYTES_PER_FRAME
  );
}

export function transcodedTotalSize(durationSeconds: number): number {
  return WAV_HEADER_SIZE + pcmBytesForDuration(durationSeconds);
}

export function buildWavHeader(pcmBytes: number): Buffer {
  const buf = Buffer.alloc(WAV_HEADER_SIZE);
  buf.write("RIFF", 0, "ascii");
  buf.writeUInt32LE(36 + pcmBytes, 4);
  buf.write("WAVE", 8, "ascii");
  buf.write("fmt ", 12, "ascii");
  buf.writeUInt32LE(16, 16); // fmt chunk body size
  buf.writeUInt16LE(1, 20); // PCM format
  buf.writeUInt16LE(CHANNELS, 22);
  buf.writeUInt32LE(SAMPLE_RATE, 24);
  buf.writeUInt32LE(PCM_BYTES_PER_SECOND, 28);
  buf.writeUInt16LE(BYTES_PER_FRAME, 32);
  buf.writeUInt16LE(BYTES_PER_SAMPLE * 8, 34);
  buf.write("data", 36, "ascii");
  buf.writeUInt32LE(pcmBytes, 40);
  return buf;
}

function attachFfmpegLogging(
  proc: ReturnType<typeof spawn>,
  filePath: string,
): void {
  proc.stderr!.on("data", (chunk: Buffer) => {
    process.stderr.write(`[ffmpeg ${filePath}] ${chunk}`);
  });
  proc.on("error", (err) => {
    console.error(`[ffmpeg spawn error] ${filePath}:`, err);
  });
  proc.on("exit", (code, signal) => {
    if (code !== 0 && code !== null) {
      console.error(`[ffmpeg exit ${code}] ${filePath} signal=${signal}`);
    }
  });
}

const FFMPEG_QUIET = ["-nostdin", "-hide_banner", "-loglevel", "error"];

// Spawns ffmpeg with input-side seek and outputs raw PCM s16le.
function spawnPcmDecoder(
  filePath: string,
  seekSeconds: number,
): ReturnType<typeof spawn> {
  const args = [...FFMPEG_QUIET];
  if (seekSeconds > 0) args.push("-ss", seekSeconds.toFixed(6));
  args.push(
    "-i",
    filePath,
    "-f",
    "s16le",
    "-acodec",
    "pcm_s16le",
    "-ar",
    String(SAMPLE_RATE),
    "-ac",
    String(CHANNELS),
    "pipe:1",
  );
  const proc = spawn(FFMPEG_PATH, args, {
    stdio: ["ignore", "pipe", "pipe"],
  });
  attachFfmpegLogging(proc, filePath);
  return proc;
}

// Returns up to pcmLengthBytes of raw PCM starting at pcmOffsetBytes.
// Pads with zero-filled silence if the source ends before the requested
// length, so callers can honor a precomputed Content-Length.
function pcmRangeStream(
  filePath: string,
  pcmOffsetBytes: number,
  pcmLengthBytes: number,
): ReadableStream {
  const seekSeconds = pcmOffsetBytes / PCM_BYTES_PER_SECOND;
  const proc = spawnPcmDecoder(filePath, seekSeconds);
  let remaining = pcmLengthBytes;
  let closed = false;
  const close = (controller: ReadableStreamDefaultController<Uint8Array>) => {
    if (closed) return;
    closed = true;
    try {
      controller.close();
    } catch {
      /* already closed */
    }
  };

  // Pads with zero-filled silence on EOF or stream error so the response
  // body always matches the promised Content-Length. Otherwise Chromium
  // reports PIPELINE_ERROR_READ on short reads.
  const finish = (controller: ReadableStreamDefaultController<Uint8Array>) => {
    if (closed) return;
    if (remaining > 0) {
      controller.enqueue(new Uint8Array(remaining));
      remaining = 0;
    }
    close(controller);
  };

  // Backpressure: ffmpeg can dump tens of MB into the pipe in <100ms.
  // Without flow control, the entire body queues into the ReadableStream
  // buffer and Electron's protocol pipe sees close() before downstream
  // drains, surfacing as Chromium PIPELINE_ERROR_READ. We pause stdout
  // when the controller's queue is full and resume from `pull`.
  return new ReadableStream<Uint8Array>({
    start(controller) {
      proc.stdout!.on("data", (chunk: Buffer) => {
        if (closed || remaining <= 0) return;
        const slice =
          chunk.length > remaining ? chunk.subarray(0, remaining) : chunk;
        controller.enqueue(new Uint8Array(slice));
        remaining -= slice.length;
        if (remaining <= 0) {
          close(controller);
          proc.kill("SIGTERM");
          return;
        }
        if ((controller.desiredSize ?? 0) <= 0) {
          proc.stdout!.pause();
        }
      });
      proc.stdout!.on("end", () => finish(controller));
      proc.stdout!.on("error", (err) => {
        console.error(`[ffmpeg stdout error] ${filePath}:`, err);
        finish(controller);
      });
      proc.on("error", () => finish(controller));
      proc.on("exit", () => {
        if (!closed && remaining > 0) finish(controller);
      });
    },
    pull() {
      if (!closed && proc.stdout!.isPaused()) proc.stdout!.resume();
    },
    cancel() {
      proc.kill("SIGTERM");
    },
  });
}

// Returns a ReadableStream covering byte range [start, end] of the
// virtual transcoded WAV file (header + PCM).
export function transcodeRange(
  filePath: string,
  durationSeconds: number,
  start: number,
  end: number,
): ReadableStream {
  const pcmTotal = pcmBytesForDuration(durationSeconds);
  const header = buildWavHeader(pcmTotal);

  const headerSlice =
    start < WAV_HEADER_SIZE
      ? header.subarray(start, Math.min(WAV_HEADER_SIZE, end + 1))
      : Buffer.alloc(0);

  const pcmStart = Math.max(0, start - WAV_HEADER_SIZE);
  const pcmEnd = end - WAV_HEADER_SIZE;
  const pcmLength = pcmEnd >= 0 ? pcmEnd - pcmStart + 1 : 0;

  if (pcmLength <= 0) {
    return new ReadableStream<Uint8Array>({
      start(controller) {
        if (headerSlice.length) controller.enqueue(new Uint8Array(headerSlice));
        controller.close();
      },
    });
  }

  const pcm = pcmRangeStream(filePath, pcmStart, pcmLength);
  return new ReadableStream<Uint8Array>({
    async start(controller) {
      if (headerSlice.length) controller.enqueue(new Uint8Array(headerSlice));
      const reader = pcm.getReader();
      try {
        for (;;) {
          const { value, done } = await reader.read();
          if (done) break;
          if (value) controller.enqueue(value);
        }
        controller.close();
      } catch (err) {
        controller.error(err);
      }
    },
  });
}

export function transcodeToWav(filePath: string): ReadableStream {
  const proc = spawn(
    FFMPEG_PATH,
    [
      ...FFMPEG_QUIET,
      "-i",
      filePath,
      "-f",
      "wav",
      "-acodec",
      "pcm_s16le",
      "-ar",
      String(SAMPLE_RATE),
      "-ac",
      String(CHANNELS),
      "pipe:1",
    ],
    { stdio: ["ignore", "pipe", "pipe"] },
  );
  attachFfmpegLogging(proc, filePath);
  return Readable.toWeb(proc.stdout!) as ReadableStream;
}

File: src-tauri/src/transcode.rs
use anyhow::{Context, Result};
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};
use std::path::Path;
use std::process::{Command, Stdio};

pub const SAMPLE_RATE: u32 = 44100;
pub const CHANNELS: u16 = 2;
pub const BYTES_PER_SAMPLE: u16 = 2;
pub const BYTES_PER_FRAME: u16 = CHANNELS * BYTES_PER_SAMPLE;
pub const PCM_BYTES_PER_SECOND: u32 = SAMPLE_RATE * BYTES_PER_FRAME as u32;
pub const WAV_HEADER_SIZE: u32 = 44;

const FFMPEG_PATH: &str = "ffmpeg";
const FFMPEG_QUIET: &[&str] = &["-nostdin", "-hide_banner", "-loglevel", "error"];
const MOOV_SCAN_BYTES: usize = 64 * 1024;

pub fn pcm_bytes_for_duration(duration_s: f64) -> u64 {
    let total_bytes = duration_s * PCM_BYTES_PER_SECOND as f64;
    let frames = (total_bytes / BYTES_PER_FRAME as f64).floor() as u64;
    frames * BYTES_PER_FRAME as u64
}

pub fn transcoded_total_size(duration_s: f64) -> u64 {
    WAV_HEADER_SIZE as u64 + pcm_bytes_for_duration(duration_s)
}

pub fn build_wav_header(pcm_bytes: u32) -> Vec<u8> {
    let mut buf = Vec::with_capacity(WAV_HEADER_SIZE as usize);
    buf.extend_from_slice(b"RIFF");
    buf.extend_from_slice(&(36u32 + pcm_bytes).to_le_bytes());
    buf.extend_from_slice(b"WAVE");
    buf.extend_from_slice(b"fmt ");
    buf.extend_from_slice(&16u32.to_le_bytes());
    buf.extend_from_slice(&1u16.to_le_bytes());
    buf.extend_from_slice(&CHANNELS.to_le_bytes());
    buf.extend_from_slice(&SAMPLE_RATE.to_le_bytes());
    buf.extend_from_slice(&PCM_BYTES_PER_SECOND.to_le_bytes());
    buf.extend_from_slice(&BYTES_PER_FRAME.to_le_bytes());
    buf.extend_from_slice(&((BYTES_PER_SAMPLE * 8) as u16).to_le_bytes());
    buf.extend_from_slice(b"data");
    buf.extend_from_slice(&pcm_bytes.to_le_bytes());
    buf
}

pub fn needs_transcode(format: &str) -> bool {
    !crate::audio_formats::is_native_format(format)
}

pub fn is_mp4_faststart(file_path: &Path) -> bool {
    let Ok(mut f) = File::open(file_path) else {
        return false;
    };
    let mut buf = vec![0u8; MOOV_SCAN_BYTES];
    let bytes_read = match f.read(&mut buf) {
        Ok(n) => n,
        Err(_) => return false,
    };
    let mut offset = 0usize;
    while offset + 8 <= bytes_read {
        let size = u32::from_be_bytes(
            buf[offset..offset + 4].try_into().unwrap_or([0; 4]),
        );
        let typ = std::str::from_utf8(&buf[offset + 4..offset + 8]).unwrap_or("");
        if typ == "moov" {
            return true;
        }
        if typ == "mdat" {
            return false;
        }
        let advance: u64 = if size == 0 {
            return false;
        } else if size == 1 {
            if offset + 16 > bytes_read {
                return false;
            }
            let hi = u32::from_be_bytes(
                buf[offset + 8..offset + 12].try_into().unwrap_or([0; 4]),
            ) as u64;
            let lo = u32::from_be_bytes(
                buf[offset + 12..offset + 16].try_into().unwrap_or([0; 4]),
            ) as u64;
            (hi << 32) | lo
        } else {
            size as u64
        };
        if advance < 8 {
            return false;
        }
        offset = offset.saturating_add(advance as usize);
    }
    false
}

pub fn should_transcode(format: &str, file_path: &Path) -> bool {
    let lo = format.to_ascii_lowercase();
    if lo == "m4a" || lo == "mp4" {
        return !is_mp4_faststart(file_path);
    }
    needs_transcode(&lo)
}

pub fn transcode_range(
    file_path: &Path,
    duration_s: f64,
    start: u64,
    end: u64,
) -> Result<Vec<u8>> {
    let length = (end - start + 1) as usize;
    let mut out = Vec::with_capacity(length);

    let pcm_total = pcm_bytes_for_duration(duration_s) as u32;
    let header = build_wav_header(pcm_total);
    if start < WAV_HEADER_SIZE as u64 {
        let header_end = (WAV_HEADER_SIZE as u64).min(end + 1) as usize;
        let header_start = start as usize;
        out.extend_from_slice(&header[header_start..header_end]);
    }

    let pcm_start = start.saturating_sub(WAV_HEADER_SIZE as u64);
    let pcm_end_excl = (end + 1).saturating_sub(WAV_HEADER_SIZE as u64);
    let pcm_len = pcm_end_excl.saturating_sub(pcm_start) as usize;
    if pcm_len > 0 {
        let seek_s = pcm_start as f64 / PCM_BYTES_PER_SECOND as f64;
        let dur_s = pcm_len as f64 / PCM_BYTES_PER_SECOND as f64 + 0.05;
        let mut child = Command::new(FFMPEG_PATH)
            .args(FFMPEG_QUIET)
            .args([
                "-ss",
                &format!("{:.6}", seek_s),
                "-t",
                &format!("{:.6}", dur_s),
                "-i",
                file_path.to_str().context("non-utf8 path")?,
                "-f",
                "s16le",
                "-acodec",
                "pcm_s16le",
                "-ar",
                &SAMPLE_RATE.to_string(),
                "-ac",
                &CHANNELS.to_string(),
                "pipe:1",
            ])
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .context("spawn ffmpeg")?;
        let mut stdout = child.stdout.take().context("ffmpeg stdout missing")?;
        let mut buf = Vec::with_capacity(pcm_len);
        stdout.read_to_end(&mut buf)?;
        let _ = child.wait();
        let take = buf.len().min(pcm_len);
        out.extend_from_slice(&buf[..take]);
    }

    let needed = length.saturating_sub(out.len());
    if needed > 0 {
        out.extend(std::iter::repeat(0u8).take(needed));
    }
    Ok(out)
}

pub fn transcode_full(file_path: &Path) -> Result<Vec<u8>> {
    let mut child = Command::new(FFMPEG_PATH)
        .args(FFMPEG_QUIET)
        .args([
            "-i",
            file_path.to_str().context("non-utf8 path")?,
            "-f",
            "wav",
            "-acodec",
            "pcm_s16le",
            "-ar",
            &SAMPLE_RATE.to_string(),
            "-ac",
            &CHANNELS.to_string(),
            "pipe:1",
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .context("spawn ffmpeg")?;
    let mut stdout = child.stdout.take().context("ffmpeg stdout missing")?;
    let mut buf = Vec::new();
    stdout.read_to_end(&mut buf)?;
    let _ = child.wait();
    Ok(buf)
}

pub fn read_file_range(path: &Path, start: u64, end: u64) -> std::io::Result<Vec<u8>> {
    let mut f = File::open(path)?;
    f.seek(SeekFrom::Start(start))?;
    let len = (end - start + 1) as usize;
    let mut buf = vec![0u8; len];
    f.read_exact(&mut buf)?;
    Ok(buf)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pcm_bytes_align_to_frames() {
        let n = pcm_bytes_for_duration(1.0);
        assert_eq!(n % BYTES_PER_FRAME as u64, 0);
        assert_eq!(n, PCM_BYTES_PER_SECOND as u64);
    }

    #[test]
    fn transcoded_size_includes_header() {
        let n = transcoded_total_size(0.0);
        assert_eq!(n, WAV_HEADER_SIZE as u64);
    }

    #[test]
    fn wav_header_is_44_bytes_and_little_endian() {
        let h = build_wav_header(0);
        assert_eq!(h.len(), WAV_HEADER_SIZE as usize);
        assert_eq!(&h[0..4], b"RIFF");
        assert_eq!(&h[8..12], b"WAVE");
        assert_eq!(&h[12..16], b"fmt ");
        assert_eq!(&h[36..40], b"data");
    }
}
